File: sdk/src/main/java/com/libreshockwave/chunks/ScriptContextChunk.java
package com.libreshockwave.chunks;

import com.libreshockwave.DirectorFile;
import com.libreshockwave.format.ChunkType;
import com.libreshockwave.id.ChunkId;
import com.libreshockwave.io.BinaryReader;

import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.List;

/**
 * Script context chunk (Lctx/LctX).
 * Contains information about scripts and their handlers.
 */
public record ScriptContextChunk(
    DirectorFile file,
    ChunkId id,
    int unknown1,
    int unknown2,
    int entryCount,
    ChunkId lnamSectionId,
    int validCount,
    int flags,
    int freePtr,
    List<ScriptEntry> entries
) implements Chunk {

    @Override
    public ChunkType type() {
        return ChunkType.Lctx;
    }

    public record ScriptEntry(
        int unknown,
        ChunkId id,
        int flags
    ) {}

    public static ScriptContextChunk read(DirectorFile file, BinaryReader reader, ChunkId id, int version) {
        // Lingo scripts are ALWAYS big endian regardless of file byte order
        reader.setOrder(ByteOrder.BIG_ENDIAN);

        int unknown1 = reader.readI32();
        int unknown2 = reader.readI32();
        int entryCount = reader.readI32();
        int entryCount2 = reader.readI32();
        int entriesOffset = reader.readU16();
        reader.skip(2); // unknown2
        reader.skip(4); // unknown3
        reader.skip(4); // unknown4
        reader.skip(4); // unknown5
        ChunkId lnamSectionId = new ChunkId(Math.max(0, reader.readI32()));
        int validCount = reader.readU16();
        int flags = reader.readU16();
        int freePtr = reader.readI16();

        List<ScriptEntry> entries = new ArrayList<>();

        if (entriesOffset > 0 && entriesOffset < reader.length()) {
            reader.setPosition(entriesOffset);
            for (int i = 0; i < entryCount; i++) {
                int entryUnknown = reader.readI32();
                // Lctx entries use negative IDs (-1) for free/empty slots; clamp to 0
                ChunkId entryId = new ChunkId(Math.max(0, reader.readI32()));
                int entryFlags = reader.readU16();
                int entryLink = reader.readI16();

                entries.add(new ScriptEntry(entryUnknown, entryId, entryFlags));
            }
        }

        return new ScriptContextChunk(
            file,
            id,
            unknown1,
            unknown2,
            entryCount,
            lnamSectionId,
            validCount,
            flags,
            freePtr,
            entries
        );
    }
}

File: cpp/src/chunks/ScriptContextChunk.cpp
#include "libreshockwave/chunks/ScriptContextChunk.hpp"

#include <algorithm>
#include <cstddef>
#include <utility>

#include "libreshockwave/io/BinaryReader.hpp"

namespace libreshockwave::chunks {

ScriptContextChunk::ScriptContextChunk(const DirectorFile* file,
                                       id::ChunkId id,
                                       int unknown1,
                                       int unknown2,
                                       int entryCount,
                                       id::ChunkId lnamSectionId,
                                       int validCount,
                                       int flags,
                                       int freePtr,
                                       std::vector<ScriptEntry> entries)
    : file_(file),
      id_(id),
      unknown1_(unknown1),
      unknown2_(unknown2),
      entryCount_(entryCount),
      lnamSectionId_(lnamSectionId),
      validCount_(validCount),
      flags_(flags),
      freePtr_(freePtr),
      entries_(std::move(entries)) {}

const DirectorFile* ScriptContextChunk::file() const { return file_; }
format::ChunkType ScriptContextChunk::type() const { return format::ChunkType::Lctx; }
id::ChunkId ScriptContextChunk::id() const { return id_; }
int ScriptContextChunk::unknown1() const { return unknown1_; }
int ScriptContextChunk::unknown2() const { return unknown2_; }
int ScriptContextChunk::entryCount() const { return entryCount_; }
id::ChunkId ScriptContextChunk::lnamSectionId() const { return lnamSectionId_; }
int ScriptContextChunk::validCount() const { return validCount_; }
int ScriptContextChunk::flags() const { return flags_; }
int ScriptContextChunk::freePtr() const { return freePtr_; }
const std::vector<ScriptContextChunk::ScriptEntry>& ScriptContextChunk::entries() const { return entries_; }

ScriptContextChunk ScriptContextChunk::read(const DirectorFile* file, io::BinaryReader& reader, id::ChunkId chunkId, int version) {
    (void)version;
    const auto originalOrder = reader.order();
    reader.setOrder(io::ByteOrder::BigEndian);

    int unknown1 = 0;
    int unknown2 = 0;
    int entryCount = 0;
    id::ChunkId lnamSectionId(0);
    int validCount = 0;
    int flags = 0;
    int freePtr = 0;
    std::vector<ScriptEntry> entries;

    if (reader.bytesLeft() >= 42) {
        unknown1 = reader.readI32();
        unknown2 = reader.readI32();
        entryCount = reader.readI32();
        (void)reader.readI32();
        const int entriesOffset = reader.readU16();
        reader.skip(2);
        reader.skip(4);
        reader.skip(4);
        reader.skip(4);
        lnamSectionId = id::ChunkId(std::max(0, reader.readI32()));
        validCount = reader.readU16();
        flags = reader.readU16();
        freePtr = reader.readI16();

        if (entriesOffset > 0 && static_cast<std::size_t>(entriesOffset) < reader.length()) {
            reader.seek(static_cast<std::size_t>(entriesOffset));
            if (entryCount > 0) {
                const auto availableEntries = reader.bytesLeft() / 12;
                entries.reserve(std::min(static_cast<std::size_t>(entryCount), availableEntries));
            }
            for (int index = 0; index < entryCount && reader.bytesLeft() >= 12; ++index) {
                const int entryUnknown = reader.readI32();
                const auto entryId = id::ChunkId(std::max(0, reader.readI32()));
                const int entryFlags = reader.readU16();
                (void)reader.readI16();
                entries.push_back(ScriptEntry{entryUnknown, entryId, entryFlags});
            }
        }
    }

    reader.setOrder(originalOrder);
    return ScriptContextChunk(file, chunkId, unknown1, unknown2, entryCount, lnamSectionId, validCount, flags, freePtr,
                              std::move(entries));
}

} // namespace libreshockwave::chunks
